### PR TITLE
make shopkeeper price rules allow string

### PR DESCRIPTION
**Pull request by GuardianDll**
#### Summary
None
#### Purpose of change
Required for #73195
#### Describe the solution
change `price_no_contents` to use string
#### Describe alternatives you've considered
None
#### Testing
TBD
#### Additional context
Sorry, i don't understand any of this code, it's mostly help of mqrause that made it work, tho i still have some issues not resolved, that i can't fix on my own

### CODE REVIEW

File: src/item.h [hunk 1]
```diff
@@ -606,7 +606,7 @@ class item : public visitable
          * If `practical` is false, returns pre-Cataclysm market value,
          * otherwise returns approximate post-cataclysm value.
          */
-        int price_no_contents( bool practical, std::optional<int> price_override = std::nullopt ) const;
```
Comment (mqrause) [inline src/item.h:609]: You shouldn't remove the `= std::nullopt`, that's the default value and the reason you had to add `std::nullopt` to the function calls.

**Comment by mqrause** [comment]
The linker error probabl means you need to include `units_fwd.h` in `faction.h`.

File: src/item.cpp [hunk 2]
```diff
@@ -7035,20 +7035,21 @@ int item::price( bool practical ) const
     int res = 0;
 
     visit_items( [&res, practical]( const item * e, item * ) {
-        res += e->price_no_contents( practical );
+        res += e->price_no_contents( practical, std::nullopt );
```
Comment (mqrause) [inline src/item.cpp:7038]: You can revert this change again.

File: tests/npc_shop_cons_rates_test.cpp [hunk 3]
```diff
@@ -16,7 +16,8 @@ TEST_CASE( "npc_shop_cons_rates", "[npc][trade]" )
     }
     WHEN( "item is matched by junk threshold" ) {
         item const glass_shard( "glass_shard" );
-        REQUIRE( glass_shard.price_no_contents( true ) < units::to_cent( myrates.junk_threshold ) );
+        REQUIRE( glass_shard.price_no_contents( true,
+                                                std::nullopt ) < units::to_cent( myrates.junk_threshold ) );
```
Comment (mqrause) [inline tests/npc_shop_cons_rates_test.cpp:20]: Same here, can be reverted.

Comment (GuardianDll) [reply tests/npc_shop_cons_rates_test.cpp:20]:
```suggestion
        REQUIRE( glass_shard.price_no_contents( true ) < units::to_cent( myrates.junk_threshold ) );
```

**Comment by GuardianDll** [comment]
it erroring, it doesn't work, i don't understand the code. Clear sign it should be done by someone smarter than me
